core/hotspot/traffic_shaping.go

```diff
@@ -155,6 +159,74 @@ func (c *baseTrafficShapingController) BoundParamIndex() int {
 	return c.paramIndex
 }
 
+// ExtractArgs matches the arg from ctx based on TrafficShapingController
+// return nil if match failed.
+func (c *baseTrafficShapingController) ExtractArgs(ctx *base.EntryContext) []interface{} {
+	if c == nil {
+		return nil
+	}
+	args := make([]interface{}, 0, len(ctx.Input.Args)+len(ctx.Input.Attachments))
+
+	indexArg := c.extractArgs(ctx)
+	if indexArg != nil {
+		args = append(args, indexArg)
+	}
+	indexArg = c.extractAttachmentArgs(ctx)
+	if indexArg != nil {
+		args = append(args, indexArg)
+	}
+	return args
+}
+func (c *baseTrafficShapingController) extractArgs(ctx *base.EntryContext) interface{} {
+	args := ctx.Input.Args
+	idx := c.BoundParamIndex()
+	if idx < 0 {
+		idx = len(args) + idx
+	}
+	if idx < 0 {
+		if logging.DebugEnabled() {
+			logging.Debug("[extractArgs] The param index of hotspot traffic shaping controller is invalid",
+				"args", args, "paramIndex", c.BoundParamIndex())
+		}
+		return nil
+	}
+	if idx >= len(args) {
+		if logging.DebugEnabled() {
+			logging.Debug("[extractArgs] The argument in index doesn't exist",
+				"args", args, "paramIndex", c.BoundParamIndex())
+		}
+		return nil
+	}
+	return args[idx]
+}
+func (c *baseTrafficShapingController) extractAttachmentArgs(ctx *base.EntryContext) interface{} {
+	attachments := ctx.Input.Attachments
+
+	if attachments == nil {
+		if logging.DebugEnabled() {
+			logging.Debug("[paramKey] The attachments of ctx is nil",
+				"args", attachments, "paramIndex", c.paramKey)
```

Comment:
`"paramIndex"` it should be paramKey,   `[paramKey]` it should be [BaseTrafficShapingController extractAttachmentArgs].

Reply:
When there is no error, errgroup becomes waitgroup.
The purpose is to count concurrently. Reduce processing time.

paramIndex has been replaced by paramKey.